tokio/src/runtime/builder.rs

```diff
@@ -374,6 +382,79 @@ impl Builder {
         self
     }
 
+    /// Executes function `f` just before a thread is parked (goes idle).
+    /// `f` is called within the Tokio context, so functions like [`tokio::spawn`](crate::spawn)
+    /// can be called, and may result in this thread being unparked immediately.
+    ///
+    /// This can be used to start work only when the executor is idle, or for bookkeeping
+    /// and monitoring purposes.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tokio::runtime;
+    /// # use std::sync::atomic::{AtomicBool, Ordering};
+    ///
+    /// # pub fn main() {
+    ///
+    /// let once = AtomicBool::new(true);
+    ///
+    /// let runtime = runtime::Builder::new_multi_thread()
+    ///     .on_thread_park(move || {
+    ///         if once.swap(false, Ordering::Relaxed) {
+    ///             tokio::spawn(async { println!("thread went idle"); });
```

Comment:
This example spawns, but its unclear if spawns are actually detected here. The runtime could just go to sleep anyway and not run it until it wakes up for some reason. (The runtime should detect such spawns, but it's unclear if it does under this implementation.)

Reply:
New test uses a `Barrier` to check that everything syncs. Confirmed by making it buggy that the test hangs if spawn from `on_thread_park` doesn't work.